library/model_support.c

```diff
@@ -86,24 +88,23 @@ int model_cmp(void *lh, void *rh, type_meta *meta) {
         void **rf_addr = (void **) ((char *) rh + fm->offset);
         void *lf_ptr, *rf_ptr;
 
-        if (fm->mod != array_mod) {
-            if (fm->mod == none_mod) {
-                lf_ptr = lf_addr;
-                rf_ptr = rf_addr;
-            }
-            else if (fm->mod == ptr_mod) {
-                lf_ptr = (void *) (*lf_addr);
-                rf_ptr = (void *) (*rf_addr);
-            }
+        if (fm->mod == none_mod) {
+            lf_ptr = lf_addr;
+            rf_ptr = rf_addr;
+            rc = ftm->comparer ? ftm->comparer(lf_ptr, rf_ptr) : model_cmp(lf_ptr, rf_ptr, ftm);
+        }
+        else if (fm->mod == ptr_mod) {
+            lf_ptr = (void *) (*lf_addr);
+            rf_ptr = (void *) (*rf_addr);
+            rc = ftm->comparer ? ftm->comparer(lf_ptr, rf_ptr) : model_cmp(lf_ptr, rf_ptr, ftm);
+        }
+        else if (fm->mod == map_mod) {
+            lf_ptr = lf_addr;
+            rf_ptr = rf_addr;
 
-            if (ftm->comparer) {
-                rc = ftm->comparer(lf_ptr, rf_ptr);
-            }
-            else {
-                rc = model_cmp(lf_ptr, rf_ptr, ftm);
-            }
+            rc = model_map_compare(lf_ptr, rf_ptr, ftm);
         }
```

Comment:
might be nice to add a catchall 'else' at the end for the next time? not that there'll be a next time? :)